index.d.ts

```diff
@@ -0,0 +1,745 @@
+/**
+ * An interface that describes what structure is expected from Snapdragon's Node.
+ */
+declare interface NodeLike<T> {
+    /**
+     * Property that is always present and true in Node.
+     */
+    readonly isNode: boolean;
+
+    /**
+     * [Optional] Value property.
+     */
+    //TODO: Which one is default or legacy?
+    value?: string;
+
+    /**
+     * Value property.
+     */
+    //TODO: Which one is default or legacy?
+    val: string;
+
+    /**
+     * [Optional] Node's children.
+     */
+    nodes?: NodeLike<T>[];
+
+    /**
+     * [Optional] Parent node.
+     */
+    parent?: NodeLike<T>;
+
+    /**
+     * Type property.
+     */
+    type: string;
+
+    /**
+     * [Optional] Unshift node to children node array (`this.nodes`).
+     */
+    unshift?: (node: NodeLike<T>) => number;
+
+    /**
+     * [Optional] Unshift node to children node array (`this.nodes`).
+     */
+    //TODO: Is this a legacy? Should it be removed?
+    unshiftNode?: (node: NodeLike<T>) => number;
+
+    /**
+     * [Optional] Push node to children node array (`this.nodes`).
+     */
+    push?: (node: NodeLike<T>) => number;
+
+    /**
+     * [Optional] Push node to children node array (`this.nodes`).
+     */
+    //TODO: Is this a legacy? Should it be removed?
+    pushNode?: (node: NodeLike<T>) => number;
+
+    /**
+     * [Optional] Pop node from children node array (`this.nodes`).
+     */
+    pop?: () => NodeLike<T> | undefined;
+
+    /**
+     * [Optional] Shift node from children node array (`this.nodes`).
+     */
+    shift?: () => NodeLike<T> | undefined;
+
+    /**
+     * [Optional] Shift node from children node array (`this.nodes`).
+     */
+    //TODO: Why null? Can't we skip index check and return empty array instead like it is supposed to happen?
+    remove?: (node: NodeLike<T>) => NodeLike<T>[] | null;
+
+    /**
+     * [Optional] Returns true if type matches `*.open` pattern.
+     */
+    //TODO: Check `utils.isOpen`.
+    isOpen?: () => boolean;
+
+    /**
+     * [Optional] Returns true if type matches `*.close` pattern.
+     */
+    //TODO: Check `utils.isClose`.
+    isClose?: () => boolean;
+
+    /**
+     * [Optional] Returns true if node is block.
+     */
+    //TODO: Check `utils.isBlock`.
+    isBlock?: () => boolean;
+
+    /**
+     * [Optional] Returns true if `node` exists in `this.nodes`.
+     */
+    has?: (node: NodeLike<T>) => boolean;
+
+    /**
+     * [Optional] Getter that returns first child node.
+     */
+    //TODO: Can be undefined and null. Why not stick to undefined?
+    readonly first: NodeLike<T> | undefined | null;
+
+    /**
+     * [Optional] Getter that returns last child node.
+     */
+    //TODO: Can be undefined and null. Why not stick to undefined?
+    readonly last: NodeLike<T> | undefined | null;
```

Comment:
> Why not stick to undefined?

In my experience, `undefined` generally means something is literally "not defined", whereas `null` is a value that was intentionally set. This distinction allows implementors to more easily know if something was intentional or not. 

I'm not opposed to simplifying that to `undefined`, but perhaps in the next major.

Reply:
Oh, I totally understand that, however this (in snapdragon-node) really bugs me out:
`return this.nodes ? this.nodes[0] : null;`

Let's say that `this.node` is empty array. Since it is not undefined we will have `this.nodes[0]` returned which is undefined.

Correct solution would be:
`return this.nodes ? (this.nodes.length > 0 ? this.nodes[0] : null) : null;`
or, if we don't care about array containing actual undefined element:
`return this.nodes ? this.nodes[0] || null : null;`